src/client/datascience/baseJupyterSession.ts

```diff
@@ -104,38 +104,14 @@ export abstract class BaseJupyterSession implements IJupyterSession {
             traceInfo(`Unhandled message found: ${m.header.msg_type}`);
         };
     }
-    public dispose(): Promise<void> {
-        return this.shutdown();
+    public async dispose(): Promise<void> {
+        await this.shutdownImplementation(false);
     }
     // Abstracts for each Session type to implement
     public abstract waitForIdle(timeout: number): Promise<void>;
 
     public async shutdown(): Promise<void> {
-        this._isDisposed = true;
-        if (this.session) {
-            try {
-                traceVerbose('Shutdown session - current session');
-                await this.shutdownSession(this.session, this.statusHandler, false);
-                traceVerbose('Shutdown session - get restart session');
-                if (this.restartSessionPromise) {
-                    this.restartSessionPromise.token.cancel();
-                    const restartSession = await this.restartSessionPromise.promise;
-                    this.restartSessionPromise.token.dispose();
-                    traceVerbose('Shutdown session - shutdown restart session');
-                    await this.shutdownSession(restartSession, undefined, true);
-                }
-            } catch {
-                noop();
-            }
-            this.setSession(undefined);
-            this.restartSessionPromise = undefined;
-            this.onStatusChangedEvent.fire('dead');
-            this._disposed.fire();
-            this._disposed.dispose();
-            this.onStatusChangedEvent.dispose();
-        }
-        disposeAllDisposables(this.disposables);
-        traceVerbose('Shutdown session -- complete');
+        await this.shutdownImplementation(true);
```

Comment:
New method with an argument to ensure remote kernels are shutdown when calling shutdown method